src/shared/telemetry/recorder.ts

```diff
@@ -0,0 +1,189 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as telemetry from './telemetry'
+import { getLogger, Logger } from '../logger'
+import { AsyncLocalStorage } from 'async_hooks'
+
+interface Metric<T> {
+    start(): ActiveMetric<T>
+    record(data: T): void
+}
+
+type ActiveMetric<T = unknown> = TelemetryRecorder<Required<T>> & Has<T> & { stop(this: T): void }
+
+interface Has<T> {
+    has<U, K extends keyof T>(this: U, ...props: K[]): this is U & { readonly [P in K]-?: T[P] }
+}
+
+interface TelemetryEvent<T = unknown> {
+    readonly name: string
+    readonly source: string
+    readonly value: T
+    readonly timestamp: Date // could use `perf_hooks`
+}
+
+interface RecordFunction<T, K extends keyof T> {
+    <U>(this: U, value: T[K]): U & { readonly [P in K]: T[P] }
+}
+
+type TelemetryRecorder<T> = {
+    readonly [P in keyof T as `record${Capitalize<P & string>}`]: RecordFunction<T, P>
+}
+
+// This works fine for prototyping, but since we already codegen telemetry, we should generate
+// the static calls as well to avoid `Proxy`
+
+function createRecorder<T, U extends Record<string, any>>(
+    inner: U,
+    source: string,
+    queue: TelemetryEvent[]
+): U & TelemetryRecorder<T> {
+    function uncapitalize(s: string): string {
+        return `${s[0].toLowerCase()}${s.slice(1)}`
+    }
+
+    const proxy = new Proxy(inner, {
+        get: (target, prop, receiver) => {
+            if (typeof prop !== 'string') {
+                throw new TypeError(`Invalid telemetry recorder key: ${String(prop)}`)
+            }
+
+            const name = uncapitalize(prop.replace('record', ''))
+
+            if (!prop.startsWith('record')) {
+                // We should find the last event with the same name but this works for now
+                return Reflect.get(target, prop, receiver) ?? queue.find(e => e.name === name)
+            }
+
+            return <U>(value: U) => {
+                queue.push({ value, source, name, timestamp: new Date() })
+                Reflect.set(target, name, value, receiver)
+
+                return proxy
+            }
+        },
+    }) as U & TelemetryRecorder<T>
+
+    return proxy
+}
+
+function startMetric<T>(id: string, record: (data: T) => void): ActiveMetric<T> {
+    const queue = storage.getStore()?.telemetry.queue ?? []
+    const createTime = new Date()
+
+    function stop() {
+        const result: Record<string, any> = { createTime }
+        result.duration = Date.now() - createTime.getTime()
```

Comment:
Even though I'm now adding 'duration' to _all_ metrics, it won't actually show up unless the metric has it defined. This problem is because our codegen filters everything out. Could call `TelemetryService.record` directly but we need the metric name/id. So that's why the codegen needs to be changed.